db/migration/migration.go

```diff
@@ -24,10 +24,193 @@ import (
 	"os"
 	"path"
+	"path/filepath"
+	"plugin"
+	"crypto/sha256"
+	"io"
+	"io/ioutil"
+	"os/exec"
+)
+
+const (
+	PLUGIN_SO_NAME = "migration_plugins_go.so"
+	PLUGIN_SHA_NAME = "migration_plugins_go.sha"
 )
 
 var logger = log.NewLogger()
 
+func collectRuntimeMigrations(migrationsPath string) (sources []string, sha string, err error) {
+	shaGen := sha256.New()
+
+	if err = filepath.Walk(migrationsPath,
+		func(walkPath string, walkInfo os.FileInfo, walkErr error) error {
+			if err != nil {
+				logger.Warning("migration: walk func error: %s", walkErr)
+				return walkErr
+			}
+
+			if walkInfo == nil {
+				return nil
+			}
+
+			if walkInfo.IsDir() {
+				return nil
+			}
+
+			if filepath.Ext(walkInfo.Name()) == ".go" {
+				sources = append(sources, walkPath)
+
+				fileHandle, fileErr := os.Open(walkPath)
+
+				if fileErr != nil {
+					return fileErr
+				}
+
+				defer fileHandle.Close()
+
+				if _, copyErr := io.Copy(shaGen, fileHandle); copyErr != nil {
+					logger.Error("migration: failed to sha file: %s", copyErr.Error())
+					return copyErr
+				}
+			}
+
+			return nil
+		},
+	); err != nil {
+		logger.Warning("migration: walk error: %s", err)
+		return
+	}
+
+	sha = fmt.Sprintf("%x", shaGen.Sum(nil))
+	return
+}
+
+func fileExists(fileName string) bool {
+	if _, err := os.Stat(fileName); err != nil {
+		if os.IsNotExist(err) {
+			return false
+		}
+	}
+	return true
+}
+
+func readRuntimeMigrationsSha(migrationsPath string) (sha string, err error) {
+	migrationsSha := migrationsPath + "/" + PLUGIN_SHA_NAME
+	if !fileExists(migrationsSha) {
+		return "", nil
+	}
+
+	fileHandle, fileErr := os.Open(migrationsSha)
+
+	if fileErr != nil {
+		logger.Error("migration: failed to open sha file for reading: %s", fileErr.Error())
+		err = fileErr
+		return
+	}
+
+	defer fileHandle.Close()
+
+	shaBytes, readErr := ioutil.ReadAll(fileHandle)
+
+	if readErr != nil {
+		logger.Error("migration: failed to read existing sha file: %s", readErr.Error())
+		err = readErr
+		return
+	}
+
+	sha = string(shaBytes)
+	return
+}
+
+func writeRuntimeMigrationsSha(migrationsPath, sha string) error {
+	fileHandle, fileErr := os.Create(migrationsPath + "/" + PLUGIN_SHA_NAME)
+
+	if fileErr != nil {
+		logger.Error("migration: failed to open sha file for writing: %s", fileErr.Error())
+		return fileErr
+	}
+
+	defer fileHandle.Close()
+
+	fileHandle.WriteString(sha)
+	return nil
+}
+
+func compileAndLoadRuntimeMigrations(runtimeSources []string, pluginName string) error {
+	os.Remove(pluginName)
+
+	logger.Info("migration: compile plugin: %s", pluginName)
+
+	commandOutput, commandErr := exec.Command("go",
+		append([]string{
+			"build",
+			"-buildmode=plugin",
+			"-o", pluginName,
+		}, runtimeSources...)...).CombinedOutput()
+
+	if commandErr != nil {
+		logger.Error("migration: failed to build runtime migrations plugin: %s; error: %s",
+			commandErr.Error(), commandOutput)
+		return commandErr
+	}
+
+	_, openErr := plugin.Open(pluginName)
```

Comment:
It is difficult to achieve since each go migration register its global methods in goose. There is no deregister handler, so goose would have to be patched. No need for such effort for so little memory savings.